README.md
[![Stories in Ready](https://badge.waffle.io/OpenBCI/OpenBCI_NodeJS.png?label=ready&title=Ready)](https://waffle.io/OpenBCI/OpenBCI_NodeJS)
[![Join the chat at https://gitter.im/OpenBCI/OpenBCI_NodeJS](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/OpenBCI/OpenBCI_NodeJS?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![Build Status](https://travis-ci.org/OpenBCI/OpenBCI_NodeJS.svg?branch=master)](https://travis-ci.org/OpenBCI/OpenBCI_NodeJS)
[![codecov](https://codecov.io/gh/OpenBCI/OpenBCI_NodeJS/branch/master/graph/badge.svg)](https://codecov.io/gh/OpenBCI/OpenBCI_NodeJS)
[![Dependency Status](https://david-dm.org/OpenBCI/OpenBCI_NodeJS.svg)](https://david-dm.org/OpenBCI/OpenBCI_NodeJS)
[![npm](https://img.shields.io/npm/dm/openbci.svg?maxAge=2592000)](http://npmjs.com/package/openbci)
[![js-semistandard-style](https://img.shields.io/badge/code%20style-semistandard-brightgreen.svg?style=flat-square)](https://github.com/Flet/semistandard)

# OpenBCI Node.js SDK

A Node.js module for OpenBCI ~ written with love by [Push The World!](http://www.pushtheworldllc.com)

We are proud to support all functionality of the OpenBCI 8 and 16 Channel boards and are actively developing and maintaining this module.

The purpose of this module is to **get connected** and **start streaming** as fast as possible.

### Table of Contents:
---

1. [Installation](#install)
2. [TL;DR](#tldr)
3. [About](#About)
4. [General Overview](#general-overview)
5. [SDK Reference Guide](#sdk-reference-guide)
  1. [Constructor](#constructor)
  2. [Methods](#method)
  3. [Events](#event)
  4. [Constants](#constants)
6. [Interfacing With Other Tools](#interfacing-with-other-tools)
7. [Developing](#developing)
8. [Testing](#developing-testing)
9. [Contribute](#contribute)
10. [License](#license)
11. [Roadmap](#roadmap)

### <a name="install"></a> Installation:
```
npm install openbci
```
### <a name="tldr"></a> TL;DR:
Get connected and start streaming right now

```js
var OpenBCIBoard = require('openbci').OpenBCIBoard;
var ourBoard = new OpenBCIBoard();
ourBoard.connect(portName) // Port name is a serial port name, see `.listPorts()`
    .then(function() {
        ourBoard.on('ready',function() {
            ourBoard.streamStart();
            ourBoard.on('sample',function(sample) {
                /** Work with sample */
                for (var i = 0; i < ourBoard.numberOfChannels(); i++) {
                    console.log("Channel " + (i + 1) + ": " + sample.channelData[i].toFixed(8) + " Volts.");
                    // prints to the console
                    //  "Channel 1: 0.00001987 Volts."
                    //  "Channel 2: 0.00002255 Volts."
                    //  ...
                    //  "Channel 8: -0.00001875 Volts."
                }
            });
        });
})
```

### <a name="about"></a> About:
Want to know if the module really works? Check out some projects and organizations using it:

* [_OpenEXP_](https://github.com/openexp/OpenEXP): an open-source desktop app for running experiments and collecting behavioral and physiological data.
* [_Thinker_](http://www.pushtheworldllc.com/#!thinker/uc1fn): a project building the world's first brainwave-word database.
* [_NeuroJS_](https://github.com/NeuroJS): a community dedicated to Neuroscience research using JavaScript, they have several great examples.

Still not satisfied it works?? Check out this [detailed report](http://s132342840.onlinehome.us/pushtheworld/files/voltageVerificationTestPlanAndResults.pdf) that scientifically validates the output voltages of this module.

How are you still doubting and not using this already? Fine, go look at some of the [700 **_automatic_** tests](https://codecov.io/gh/OpenBCI/OpenBCI_NodeJS) written for it!

### <a name="general-overview"></a> General Overview:

Initialization
--------------

Initializing the board:

```js
var OpenBCIBoard = require('openbci');
var ourBoard = new OpenBCIBoard.OpenBCIBoard();
```

For initializing with options, such as verbose print outs:

```js
var OpenBCIBoard = require('openbci').OpenBCIBoard;
var ourBoard = new OpenBCIBoard({
    verbose: true
});
```

Or if you don't have a board and want to use synthetic data:

```js
var OpenBCIBoard = require('openbci').OpenBCIBoard;
var ourBoard = new OpenBCIBoard({
    simulate: true
});
```

Another useful way to start the simulator:
```js
var openBCIBoard = require('openbci');
var k = openBCIBoard.OpenBCIConstants;
var ourBoard = openBCIBoard.OpenBCIBoard();
ourBoard.connect(k.OBCISimulatorPortName) // This will set `simulate` to true
    .then(function(boardSerial) {
        ourBoard.on('ready',function() {
            /** Start streaming, reading registers, what ever your heart desires  */
        });
    }).catch(function(err) {
        /** Handle connection errors */
    });
```


'ready' event
------------

You MUST wait for the 'ready' event to be emitted before streaming/talking with the board. The ready happens asynchronously
so installing the 'sample' listener and writing before the ready event might result in... nothing at all.

```js
var OpenBCIBoard = require('openbci').OpenBCIBoard;
var ourBoard = new OpenBCIBoard();
ourBoard.connect(portName).then(function(boardSerial) {
    ourBoard.on('ready',function() {
        /** Start streaming, reading registers, what ever your heart desires  */
    });
}).catch(function(err) {
    /** Handle connection errors */
});            
```

Sample properties:
------------------
* `startByte` (`Number` should be `0xA0`)
* `sampleNumber` (a `Number` between 0-255)
* `channelData` (channel data indexed at 0 filled with floating point `Numbers` in Volts)
* `accelData` (`Array` with X, Y, Z accelerometer values when new data available)
* `auxData` (`Buffer` filled with either 2 bytes (if time synced) or 6 bytes (not time synced))
* `stopByte` (`Number` should be `0xCx` where x is 0-15 in hex)
* `boardTime` (`Number` the raw board time)
* `timeStamp` (`Number` the `boardTime` plus the NTP calculated offset)

The power of this module is in using the sample emitter, to be provided with samples to do with as you wish.

You can also start the simulator by sending [`.connect(portName)`](#method-connect) with `portName` equal to [`'OpenBCISimulator'`](#constants-obcisimulatorportname).

To get a ['sample'](#event-sample) event, you need to:
-------------------------------------
1. Call [`.connect(serialPortName)`](#method-connect)
2. Install the ['ready'](#event-ready) event emitter on resolved promise
3. In callback for ['ready'](#event-ready) emitter, call [`streamStart()`](#method-stream-start)
4. Install the ['sample'](#event-sample) event emitter
```js
var OpenBCIBoard = require('openbci').OpenBCIBoard;
var ourBoard = new OpenBCIBoard();
ourBoard.connect(portName).then(function() {
    ourBoard.on('ready',function() {
        ourBoard.streamStart();
        ourBoard.on('sample',function(sample) {
            /** Work with sample */
        });
    });
}).catch(function(err) {
    /** Handle connection errors */
});            
```
Close the connection with [`.streamStop()`](#method-stream-stop) and disconnect with [`.disconnect()`](#method-disconnect)
```js
var ourBoard = new require('openbci').OpenBCIBoard();
ourBoard.streamStop().then(ourBoard.disconnect());
```

Time Syncing
------------
You must be using OpenBCI firmware version 2 in order to do time syncing. After you [`.connect()`](#method-connect) and send a [`.softReset()`](#method-soft-reset), you can call [`.usingVersionTwoFirmware()`](#method-using-version-two-firmware) to get a boolean response as to if you are using `v1` or `v2`.

Now using firmware `v2`, the fun begins! We synchronize the Board's clock with the module's time. In firmware `v2` we leverage samples with time stamps and _ACKs_ from the Dongle to form a time synchronization strategy. Time syncing has been verified to +/- 4ms and a test report is on the way. We are still working on the synchronize of this module and an NTP server, this is an open call for any NTP experts out there! With a global NTP server you could use several different devices and all sync to the same time server. That way you can really do some serious cloud computing!

Keep your resync interval above 50ms. While it's important to resync every couple minutes due to drifting of clocks, please do not try to sync without getting the last sync event! We can only support one sync operation at a time!

Using local computer time:
```js
var OpenBCIBoard = require('openbci').OpenBCIBoard,
    ourBoard = new OpenBCIBoard({
        verbose:true
    });

const resyncPeriodMin = 5; // re sync every five minutes
const secondsInMinute = 60;
var sampleRate = k.OBCISampleRate250; // Default to 250, ALWAYS verify with a call to `.sampleRate()` after 'ready' event!
var timeSyncPossible = false;

// Call to connect
ourBoard.connect(portName).then(() => {
    ourBoard.on('ready',() => {
        // Get the sample rate after 'ready'
        sampleRate = ourBoard.sampleRate();
        // Find out if you can even time sync, you must be using v2 and this is only accurate after a `.softReset()` call which is called internally on `.connect()`. We parse the `.softReset()` response for the presence of firmware version 2 properties.
        timeSyncPossible = ourBoard.usingVersionTwoFirmware();

        ourBoard.streamStart()
            .then(() => {
                /** Start streaming command sent to board. */
            })
            .catch(err => {
                console.log(`stream start: ${err}`);
            })
    });

    // PTW recommends sample driven  
    ourBoard.on('sample',sample => {
        // Resynchronize every every 5 minutes
        if (sample._count % (sampleRate * resyncPeriodMin * secondsInMinute) === 0) {
            ourBoard.syncClocksFull()
                .then(syncObj => {
                    // Sync was successful
                    if (syncObj.valid) {
                        // Log the object to check it out!
                        console.log(`syncObj`,syncObj);

                    // Sync was not successful
                    } else {
                        // Retry it
                        console.log(`Was not able to sync, please retry?`);
                    }
                });
        }

        if (sample.timeStamp) { // true after the first sync
            console.log(`NTP Time Stamp ${sample.timeStamp}`);
        }

    });
})
.catch(err => {
    console.log(`connect: ${err}`);
});
```

Auto-finding boards
-------------------
You must have the OpenBCI board connected to the PC before trying to automatically find it.

If a port is not automatically found, then call [`.listPorts()`](#method-list-ports) to get a list of all serial ports this would be a good place to present a drop down picker list to the user, so they may manually select the serial port name.

```js
var OpenBCIBoard = require('openbci').OpenBCIBoard;
var ourBoard = new OpenBCIBoard();
ourBoard.autoFindOpenBCIBoard().then(portName => {
    if(portName) {
        /**
        * Connect to the board with portName
        * i.e. ourBoard.connect(portName).....
        */
    } else {
        /**Unable to auto find OpenBCI board*/
    }
});
```

Note: [`.autoFindOpenBCIBoard()`](#method-auto-find-open-bci-board) will return the name of the Simulator if you instantiate with option `simulate: true`.

Auto Test - (Using impedance to determine signal quality)
---------------------------------------------------------
Measuring impedance is a vital tool in ensuring great data is collected.

**_IMPORTANT!_** Measuring impedance takes time, so *only test what you must*

Your OpenBCI board will have electrodes hooked up to either a P input, N input or in some cases both inputs.

To test specific inputs of channels:

1. Connect to board.
2. Start streaming.
3. Install the ['impedanceArray'](#event-impedance-array) event
4. Call [`.impedanceTestChannels()`](#method-impedance-test-channels) with your configuration array

A configuration array looks like, for an 8 channel board, `['-','N','n','p','P','-','b','b']`

Where there are the same number of elements as channels and each element can be either:

* `p` or `P` (only test P input)
* `n` or `N` (only test N input)
* `b` or `B` (test both inputs) (takes 66% longer to run then previous two `p` or `n`)
* `-` (ignore channel)

Without further ado, here is an example:
```js
var OpenBCIBoard = require('openbci').OpenBCIBoard;
var ourBoard = new OpenBCIBoard();
ourBoard.connect(portName).then(function(boardSerial) {
    ourBoard.on('ready',function() {
        ourBoard.streamStart();
        ourBoard.once('impedanceArray', impedanceArray => {
            /** Work with impedance Array */
        });
        ourBoard.impedanceTestChannels(['n','N','n','p','P','p','b','B']).catch(err => console.log(err));
    });
}).catch(function(err) {
    /** Handle connection errors */
});
```

But wait! What is this `impedanceArray`? An Array of Objects, for each object:
```js
[{
    channel: 1,
    P: {
        raw: -1,
        text: 'init'
    },
    N: {
        raw: -1,
        text: 'init'
    }
},
{
    // Continues for each channel up to the amount of channels on board (8 or 16)
},...];
```

Where:

* *channel* is the channel number (`impedanceArray[0]` is channel 1, `impedanceArray[6]` is channel 7)
* *P* is the P input data (Note: P is capitalized)
  * *raw* is an impedance value resulting from the Goertzel algorithm.
  * *text* is a text interpretation of the `average`
    * **Good** impedance is < 5k Ohms
    * **Ok** impedance is 5 to 10k Ohms
    * **Bad** impedance is > 10k Ohms
    * **None** impedance is > 1M Ohms
* *N* is the N input data (Note: N is capitalized) (see above for what N object consists of)

To run an impedance test on all inputs, one channel at a time:

1. Connect to board
2. Start streaming
3. Install the ['impedanceArray'](#event-impedance-array)
4. Call [`.impedanceTestAllChannels()`](#method-impedance-test-all-channels)

**Note: Takes up to 5 seconds to start measuring impedances. There is an unknown number of samples taken. Not always 60!**

For example:

```js
var OpenBCIBoard = require('openbci').OpenBCIBoard;
var ourBoard = new OpenBCIBoard();
ourBoard.connect(portName).then(function(boardSerial) {
    ourBoard.streamStart();
    ourBoard.on('impedanceArray', impedanceArray => {
        /** Work with impedance */
    });
    ourBoard.impedanceTestAllChannels();
}
```

See Reference Guide for a complete list of impedance tests.

# <a name="sdk-reference-guide"></a> SDK Reference Guide:
---------------
## <a name="constructor"></a> Constructor:

### <a name="constructor-openbciboard"></a> OpenBCIBoard (options)

Create new instance of an OpenBCI board.

**_options (optional)_**

Board optional configurations.

* `baudRate` {Number} - Baud Rate, defaults to 115200. Manipulating this is allowed if firmware on board has been previously configured.

changelog.md
# 1.4.1

### New Features
* New API function `.isSimulating()` to check if module is in simulate mode.

### Bug Fixes
* Fixes bug where extra data after EOT (`$$$`) was dumped by preserving the poriton after the EOT for further decomposition.
* Fixes bug where any calls to channel set would actually break the openBCISample code as the channelSettingsArray contained an undefined.

# 1.4.0

### New Features

* Three new initialization options: `simulatorFragmentation`, `simulatorBufferSize`, and `simulatorLatencyTimer`.  Together, these enable a more _realistic_ serial port simulation, mimicking different potential user computer systems.
* New option `debug` gives a live dump of serial traffic on the console if enabled
* New API function `.isConnected()` to check if communications are active.
* New API function `.isStreaming()` to check if samples are coming from the board.

### Enhancements

* Implement and adapt semi-standard code style. Closes #83
* autoFindOpenBCIBoard now notices and uses the stock dongle on Linux
* 'synced' object now has `error` property, null on good syncs, error description on bad syncs.

### Breaking Changes

* The setting for simulatorInjectLineNoise has changed from `None` to `none`
* connect() will now fail if already connected
* The constructor will throw an error now if an invalid option is passed
* The `.connected` property has been removed, replaced by `.isConnected()`. Removed from docs.
* The `.streaming` property has been removed, replaced by `.isStreaming()`. Removed from docs.
* An error event will be emitted if sntp fails to initialize on construction
* The simulator will no longer communicate when disconnected
* Promises returned by writes will now only resolve after the write has been sent

### Bug Fixes

* Fixed bug where early packet fragments were dropped after board reset
* Fixed bug where time sync replies that began a buffered chunk were ignored
* Fixed bug where simulator would output wrong version in its reset message
* Fixed bug where resources were not cleaned up if connect was called twice
* Fixed bug where serial data was written after disconnection
* Fixed bug where unexpected disconnection was not detected
* Fixed bug where promises could lead to out of order packet processing.

# 1.3.3

### New Features

* Add `timeOffsetMaster` to object emitted when bad time sync.

### Bug Fixes

* Fixed log statement on impedance setting function
* Remove event emitter with time sync on reject of sync clock full

# 1.3.2

### Enhancements

* Added master time offset `timeOffsetMaster` to `syncObj` which is a running average across sync attempts.

# 1.3.1

### Bug Fixes

* Fixed bug where `connected` and `streaming` were not set in constructor

# 1.3.0

### New Features

* Add dropped packet detection, new event `droppedPacket` can be added to get an array of dropped packet numbers in the case of the dropped packet event.

# 1.2.3

### Enhancements

* Add table of contents to read me
* Reduce size of repo by removing impedance test report

# 1.2.2

### Enhancements

* Upgrade serialport to 4.x

# 1.2.1

### Bug Fixes

* Fixed bug where set channel function allowed for channel 0 to be set. Cannot set system to channel 0; lower limit is 1.

# 1.2.0

### New Features

* Add tutorial/sample code for interfacing the module with lab streaming layer.

### Breaking Changes

* Fixed time synced accel to work OpenBCI_32bit_Library release candidate 5 and newer.

# 1.1.0

### New Features

* Add function `.time()` which should be used in time syncing
* Add function `.syncClocksFull()` which should be used for immediate consecutive time syncs
* Synced object can be emitted on `synced` event. Check `valid` property for if the sync was done
* Add detailed description of object returned on `synced` event to README.md

### Breaking Changes

* Changed option named `timeSync` to `sntpTimeSync`
* Removed function called `.sntpNow()` because it was replaced by `.time()`

### Bug Fixes

* Time sync working
* Module could not work with local time

# 1.0.1

### New Features

* Add time sync tutorial in `README.md`

### Bug Fixes

* Fixed bug in simulator that lead to samples being all zeros.
* Fixed time sync sent confirmation bug that led to bad time values.

# 1.0.0

The second major release for the OpenBCI Node.js SDK brings major changes, improvements and stability, on top of a push to increase automated test coverage.

### New Features

* NPM Module now called `openbci` instead of `openbci-sdk`
* Firmware version 2.0 support.
* Clock synchronization between OpenBCI board and this driver
* `.info` property which contains verified board information parsed from the OpenBCI's `softReset()` response.
* `eot` event that is emitted when a user sends a command that results in an EOT ("$$$") being sent from the board
* Daisy (16 channel) support
* Simulator overhaul, it completely mocks the board. Can now simulate board failure, where the board stops talking to the dongle. Can also mock a serial port failure.
* `error` and `close` events from serialport now emtted events users can subscribe to.

### Breaking Changes

* NPM package is not called `openbci-sdk` anymore, now called `openbci`
* Accelerometer data now goes into `.accelData` array instead of `.auxData` array.
* In openBCISample.js
  * `parseRawPacket()` is now called `parseRawPacketStandard()`
* `ready` event only triggered after soft reset. `eot` event emitted in all other conditions resulting in the board sending EOT ("$$$")
* Must use camel case on the OpenBCISimulator object.
* Renamed constructor options for readability:
  * `simulatorAlpha` to `simulatorInjectAlpha`
  * `simulatorLineNoise` to `simulatorInjectLineNoise`
* `connect()` no longer rejects on `close` or `error` event from the serialport.

# 0.3.9

### Enhancements

* Add tests for parsing raw packets

### Bug Fixes

* Removed `got here` log from `.streamStart()`
* Validate stop byte before emitting `rawDataPacket`

# 0.3.8

### Bug Fixes

openBCISimulator.js
'use strict';

var EventEmitter = require('events').EventEmitter;
var util = require('util');
var stream = require('stream');

var openBCISample = require('./openBCISample');
var k = openBCISample.k;
var now = require('performance-now');


function OpenBCISimulatorFactory() {
    var factory = this;

    var _options = {
        accel: true,
        alpha: true,
        boardFailure:false,
        daisy: false,
        drift: 0,
        firmwareVersion: k.OBCIFirmwareV1,
        lineNoise: '60Hz',
        sampleRate: 250,
        serialPortFailure:false,
        verbose: false
    };

    function OpenBCISimulator(portName, options) {
        options = (typeof options !== 'function') && options || {};
        var opts = {};

        stream.Stream.call(this);

        /** Configuring Options */
        if (options.accel === false) {
            opts.accel = false;
        } else {
            opts.accel = _options.accel;
        }
        if (options.alpha === false) {
            opts.alpha = false;
        } else {
            opts.alpha = _options.alpha;
        }
        opts.boardFailure = options.boardFailure || _options.boardFailure;
        opts.daisy = options.daisy || _options.daisy;
        opts.drift = options.drift || _options.drift;
        opts.firmwareVersion = options.firmwareVersion || _options.firmwareVersion;
        opts.lineNoise = options.lineNoise || _options.lineNoise;
        if (options.sampleRate) {
            opts.sampleRate = options.sampleRate;
        } else {
            opts.sampleRate = k.OBCISampleRate250;
        }
        opts.serialPortFailure = options.serialPortFailure || _options.serialPortFailure;
        opts.verbose = options.verbose || _options.verbose;

        this.options = opts;

        // Bools
        this.connected = false;
        this.sd = {
            active:false,
            startTime: 0
        };
        this.streaming = false;
        this.synced = false;
        this.sendSyncSetPacket = false;
        // Buffers
        this.buffer = new Buffer(500);
        // Numbers
        this.channelNumber = 1;
        this.hostChannelNumber = this.channelNumber;
        this.pollTime = 80;
        this.sampleNumber = -1; // So the first sample is 0
        // Objects
        this.sampleGenerator = openBCISample.randomSample(k.OBCINumberOfChannelsDefault, this.options.sampleRate, this.options.alpha, this.options.lineNoise);
        this.time = {
            current: 0,
            start: now(),
            loop: null
        };
        // Strings
        this.portName = portName || k.OBCISimulatorPortName;

        // Call 'open'
        if (this.options.verbose) console.log(`Port name: ${portName}`);
        setTimeout(() => {
            this.emit('open');
            this.connected = true;
        }, 200);

    }

    // This allows us to use the emitter class freely outside of the module
    util.inherits(OpenBCISimulator, stream.Stream);

    OpenBCISimulator.prototype.flush = function() {
        this.buffer.fill(0);
        //if (this.options.verbose) console.log('flushed');
    };

    OpenBCISimulator.prototype.write = function(data,callback) {
        switch (data[0]) {
            case k.OBCIRadioKey:
                this._processPrivateRadioMessage(data);
                break;
            case k.OBCIStreamStart:
                if (!this.stream) this._startStream();
                this.streaming = true;
                break;
            case k.OBCIStreamStop:
                if (this.stream) clearInterval(this.stream); // Stops the stream
                this.streaming = false;
                break;
            case k.OBCIMiscSoftReset:
                if (this.stream) clearInterval(this.stream);
                this.streaming = false;
                this.emit('data', new Buffer(`OpenBCI V3 Simulator\nOn Board ADS1299 Device ID: 0x12345\n${this.options.daisy ? "On Daisy ADS1299 Device ID: 0xFFFFF\n" : ""}LIS3DH Device ID: 0x38422\n${this.options.firmware === k.OBCIFirmwareV2 ? "Firmware: v2.0.0\n" : ""}$$$`));
                break;
            case k.OBCISDLogForHour1:
            case k.OBCISDLogForHour2:
            case k.OBCISDLogForHour4:
            case k.OBCISDLogForHour12:
            case k.OBCISDLogForHour24:
            case k.OBCISDLogForMin5:
            case k.OBCISDLogForMin15:
            case k.OBCISDLogForMin30:
            case k.OBCISDLogForSec14:
                // If we are not streaming, then do verbose output
                if (!this.streaming) {
                    this.emit('data', new Buffer('Wiring is correct and a card is present.\nCorresponding SD file OBCI_69.TXT\n$$$'));
                }
                this.sd.active = true;
                this.sd.startTime = now();
                break;
            case k.OBCISDLogStop:
                if (!this.streaming) {
                    if (this.SDLogActive) {
                        this.emit('data', new Buffer(`Total Elapsed Time: ${now() - this.sd.startTime} ms\n`));
                        this.emit('data', new Buffer(`Max write time: ${Math.random()*500} us\n`));
                        this.emit('data', new Buffer(`Min write time: ${Math.random()*200} us\n`));
                        this.emit('data', new Buffer(`Overruns: 0\n`));
                        this._printEOT();
                    } else {
                        this.emit('data', new Buffer('No open file to close\n'));
                        this._printEOT();
                    }
                }
                this.SDLogActive = false;
                break;
            case k.OBCISyncTimeSet:
                if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
                    this.synced = true;
                    setTimeout(() => {
                        this.emit('data', new Buffer(k.OBCISyncTimeSent));
                        this._syncUp();
                    }, 10);
                }
                break;
            default:
                break;
        }

        /** Handle Callback */
        if (this.connected) {
            callback(null,'Success!');
        }
    };

    OpenBCISimulator.prototype.drain = function(callback) {
        callback();
    };

    OpenBCISimulator.prototype.close = function(callback) {

        // TODO: remove these when there is a code path allowing simulator re-use
        clearInterval(this.stream);
        this.streaming = false;

        if (this.connected) {
            this.emit('close');
        }
        this.connected = false;
        callback();
    };

    OpenBCISimulator.prototype._startStream = function() {
        var intervalInMS = 1000 / this.options.sampleRate;

        if (intervalInMS < 2) intervalInMS = 2;

        var getNewPacket = sampNumber => {
            if (this.options.accel) {
                if (this.synced) {
                    if (this.sendSyncSetPacket) {
                        this.sendSyncSetPacket = false;
                        return openBCISample.convertSampleToPacketAccelTimeSyncSet(this.sampleGenerator(sampNumber),now().toFixed(0));
                    } else {
                        return openBCISample.convertSampleToPacketAccelTimeSynced(this.sampleGenerator(sampNumber),now().toFixed(0));
                    }
                } else {
                    return openBCISample.convertSampleToPacketStandard(this.sampleGenerator(sampNumber));
                }
            } else {
                if (this.synced) {
                    if (this.sendSyncSetPacket) {
                        this.sendSyncSetPacket = false;
                        return openBCISample.convertSampleToPacketRawAuxTimeSyncSet(this.sampleGenerator(sampNumber),now().toFixed(0),new Buffer([0,0,0,0,0,0]));
                    } else {
                        return openBCISample.convertSampleToPacketRawAuxTimeSynced(this.sampleGenerator(sampNumber),now().toFixed(0),new Buffer([0,0,0,0,0,0]));
                    }
                } else {
                    return openBCISample.convertSampleToPacketRawAux(this.sampleGenerator(sampNumber),new Buffer([0,0,0,0,0,0]));
                }

            }
        };

        this.stream = setInterval(() => {
            this.emit('data', getNewPacket(this.sampleNumber));
            this.sampleNumber++;
        }, intervalInMS);
    };

    OpenBCISimulator.prototype._syncUp = function() {
        setTimeout(() => {
            this.sendSyncSetPacket = true;
        }, 12); // 3 packets later
    };

    OpenBCISimulator.prototype._printEOT = function () {
        this.emit('data', new Buffer("$$$"));
    };

    OpenBCISimulator.prototype._printFailure = function () {
        this.emit('data', new Buffer("Failure: "));
    };

    OpenBCISimulator.prototype._printSuccess = function () {
        this.emit('data', new Buffer("Success: "));
    };

    OpenBCISimulator.prototype._printValidatedCommsTimeout = function () {
        this._printFailure();
        this.emit('data', new Buffer("Communications timeout - Device failed to poll Host"));
        this._printEOT();
    };

    OpenBCISimulator.prototype._processPrivateRadioMessage = function(dataBuffer) {
        switch (dataBuffer[1]) {
            case k.OBCIRadioCmdChannelGet:
                if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
                    if (!this.options.boardFailure) {
                        this._printSuccess();
                        this.emit('data', new Buffer(`Host and Device on Channel Number ${this.channelNumber}`));
                        this.emit('data', new Buffer([this.channelNumber]));
                        this._printEOT();
                    } else if (!this.serialPortFailure) {
                        this._printFailure();
                        this.emit('data', new Buffer(`Host on Channel Number ${this.channelNumber}`));
                        this.emit('data', new Buffer([this.channelNumber]));
                        this._printEOT();
                    }
                }
                break;
            case k.OBCIRadioCmdChannelSet:
                if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
                    if (!this.options.boardFailure) {
                        if (dataBuffer[2] < k.OBCIRadioChannelMax) {
                            this.channelNumber = dataBuffer[2];
                            this.hostChannelNumber = this.channelNumber;
                            this._printSuccess();
                            this.emit('data', new Buffer(`Channel Number ${this.channelNumber}`));
                            this.emit('data', new Buffer([this.channelNumber]));
                            this._printEOT();
                        } else {
                            this._printFailure();
                            this.emit('data', new Buffer("Verify channel number is less than 25"));
                            this._printEOT();
                        }
                    } else if (!this.serialPortFailure) {
                       this._printValidatedCommsTimeout();
                    }
                }
                break;
            case k.OBCIRadioCmdChannelSetOverride:
                if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
                    if (dataBuffer[2] < k.OBCIRadioChannelMax) {
                        if (dataBuffer[2] === this.channelNumber) {
                            this.options.boardFailure = false;
                        } else {
                            this.options.boardFailure = true;
                        }
                        this.hostChannelNumber = dataBuffer[2];
                        this._printSuccess();
                        this.emit('data', new Buffer(`Host override - Channel Number ${this.hostChannelNumber}`));
                        this.emit('data', new Buffer([this.hostChannelNumber]));
                        this._printEOT();
                    } else {
                        this._printFailure();
                        this.emit('data', new Buffer("Verify channel number is less than 25"));
                        this._printEOT();
                    }
                }
                break;
            case k.OBCIRadioCmdPollTimeGet:
                if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
                    if (!this.options.boardFailure) {
                        this._printSuccess();
                        this.emit('data', new Buffer(`Poll Time ${this.pollTime}`));
                        this.emit('data', new Buffer([this.pollTime]));
                        this._printEOT();
                    } else {
                        this._printValidatedCommsTimeout();
                    }
                }
                break;
            case k.OBCIRadioCmdPollTimeSet:
                if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
                    if (!this.options.boardFailure) {
                        this.pollTime = dataBuffer[2];
                        this._printSuccess();
                        this.emit('data', new Buffer(`Poll Time ${this.pollTime}`));
                        this.emit('data', new Buffer([this.pollTime]));
                        this._printEOT();
                    } else {
                        this._printValidatedCommsTimeout();
                    }
                }
                break;
            case k.OBCIRadioCmdBaudRateSetDefault:
                if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
                    this._printSuccess();
                    this.emit('data', new Buffer("Switch your baud rate to 115200"));
                    this.emit('data', new Buffer([0x24,0x24,0x24,0xFF])); // The board really does this
                }
                break;
            case k.OBCIRadioCmdBaudRateSetFast:
                if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
                    this._printSuccess();
                    this.emit('data', new Buffer("Switch your baud rate to 230400"));
                    this.emit('data', new Buffer([0x24,0x24,0x24,0xFF])); // The board really does this
                }
                break;
            case k.OBCIRadioCmdSystemStatus:
                if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
                    if (!this.options.boardFailure) {
                        this._printSuccess();
                        this.emit('data', new Buffer("System is Up"));
                        this._printEOT();
                    } else {
                        this._printFailure();
                        this.emit('data', new Buffer("System is Down"));
                        this._printEOT();
                    }
                }
                break;
            default:
                break;
        }
    };

    factory.OpenBCISimulator = OpenBCISimulator;
}

util.inherits(OpenBCISimulatorFactory, EventEmitter);

module.exports = new OpenBCISimulatorFactory();
